### ISSUE TITLE
Support takeover for devicemanager/device-plugin

### ISSUE THREAD

**Issue by linxiulei** [kubernetes/kubernetes#127803]
### What would you like to be added?

Add support for seamless takeover by a new device-plugin for a resource without deregistering the resource.

In other words, a new device-plugin can be started and registering the resource before the old device-plugin is stopped. In this case, devicemanager should not change the values of allocatable/total count for this resource.

### Why is this needed?

Currently, devicemanager only works with one device-plugin instance for every resource. In other words, if two device-plugin are running, devicemanager may have unexpected behaviors. 

For example, when two device-plugin are running, they are both registered for the same resource. However, if one of them stops, the resource is marked as deregistered and resource counts reset to 0.

We need this support to have seamless deployment of device-plugin. After this is supported, new device-plugin can be started while the old device-plugin is also running, and devicemanager should notice the new device-plugin, as it registers more recently, to be the new takeover and gracefully deregister/disconnect the old device-plugin. With this, there will be no downtime for resources in device-plugin updates.

**Comment by ffromani** [comment]
This seems interesting. What's the usecase for this feature? Which user flow benefit from this?

In addition: how can we safely distinguish a takeover, so we avoid to report stale state from the kubelet?

**Comment by linxiulei** [comment]
The impact may be not that visible if the cluster has only a handful of nodes with resources. I didn't search for existing issues but my own observation on a large cluster of thousands of GPUs was that whenever there is an update for the device plugin pod for the cluster, a visible amount of nodes report 0 GPU counts for a short period of time.

**Comment by ffromani** [comment]
> The impact may be not that visible if the cluster has only a handful of nodes with resources. I didn't search for existing issues but my own observation on a large cluster of thousands of GPUs was that whenever there is an update for the device plugin pod for the cluster, a visible amount of nodes report 0 GPU counts for a short period of time.

yes, the counts dropping to 0 is how it is meant to work with the current code. I'm trying to understand if the counter reset to 0 has some negative effect on workload.
I can think the following: some nodes won't be used by the scheduler from a short period of time till the node publishes the updated state again. Are there more issues or this is it?

I'm asking because the takeover doesn't look trivial to implement in the current architecture, so changes must be evaluated carefully in terms of cost/benefit ration

**Comment by linxiulei** [comment]
Thanks for explaining it!

That's it. But it causes noises in monitoring as it seems like the devices are not available for a while. Seamless update would be ideal.

I took a look at the devicemanager code and may have overlooked something that'd be much more complicated than I think so I oversimplified it, but I can take a stab at it by putting up a quick PR to implement, which shouldn't be too complex.

**Comment by ffromani** [comment]
> That's it. But it causes noises in monitoring as it seems like the devices are not available for a while. Seamless update would be ideal.
> 
> I took a look at the devicemanager code and may have overlooked something that'd be much more complicated than I think so I oversimplified it, but I can take a stab at it by putting up a quick PR to implement, which shouldn't be too complex.

I'm not against changing the behavior, even though looks like a feature rather than an issue (I'll fix labels). What I'm most worried about is detecting takeover vs keeping stale counters. I guess we can discuss over the PR implementation.

**Comment by lengrongfu** [comment]
I checked your PR and it doesn't seem to solve all your problems. Your PR only reduces the probability of this problem occurring.

Because device-plugin is always deleted and then started, there will definitely be a situation where the resources are 0 during this interval.

@linxiulei

**Comment by linxiulei** [comment]
Please correct me if I'm wrong, according to https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/daemon-set-v1/#DaemonSetSpec

When updateStrategy.rollingUpdate.maxSurge is specified:

> Example: when this is set to 30%, at most 30% of the total number of nodes that should be running the daemon pod (i.e. status.desiredNumberScheduled) can have their a new pod created before the old pod is marked as deleted. The update starts by launching new pods on 30% of nodes. Once an updated pod is available (Ready for at least minReadySeconds) the old DaemonSet pod on that node is marked deleted.

My understanding on this is for those nodes there can be 2 device-plugin pods running at the same time for the period of update.

**Comment by lengrongfu** [comment]
> Please correct me if I'm wrong, according to https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/daemon-set-v1/#DaemonSetSpec
> 
> When updateStrategy.rollingUpdate.maxSurge is specified:
> 
> > Example: when this is set to 30%, at most 30% of the total number of nodes that should be running the daemon pod (i.e. status.desiredNumberScheduled) can have their a new pod created before the old pod is marked as deleted. The update starts by launching new pods on 30% of nodes. Once an updated pod is available (Ready for at least minReadySeconds) the old DaemonSet pod on that node is marked deleted.
> 
> My understanding on this is for those nodes there can be 2 device-plugin pods running at the same time for the period of update.

Thank you for your explanation. Modifying this configuration `updateStrategy.rollingUpdate.maxSurge` does work.

**Comment by SergeyKanzhelev** [comment]
BTW, it is now supported for DRA: https://github.com/kubernetes/kubernetes/pull/129832
